src/rpc/net.cpp

```diff
@@ -234,23 +234,46 @@ UniValue addnode(const JSONRPCRequest& request)
 
 UniValue disconnectnode(const JSONRPCRequest& request)
 {
-    if (request.fHelp || request.params.size() != 1)
+    if (request.fHelp || request.params.size() == 0 || request.params.size() >= 3)
         throw std::runtime_error(
-            "disconnectnode \"address\" \n"
-            "\nImmediately disconnects from the specified node.\n"
+            "disconnectnode \"address\" [nodeid]\n"
+            "\nImmediately disconnects from the specified peer node.\n"
+            "\nStrictly one out of 'address' and 'nodeid' can be provided to identify the node.\n"
+            "\nTo disconnect by nodeid, either set \"address\" to the empty string, or call using the named \"nodeid\" argument only.\n"
             "\nArguments:\n"
-            "1. \"address\"     (string, required) The IP address/port of the node\n"
+            "1. \"address\"     (string, optional) The IP address/port of the node\n"
+            "2. \"nodeid\"      (int, optional) The node ID (see getpeerinfo for node IDs)\n"
             "\nExamples:\n"
             + HelpExampleCli("disconnectnode", "\"192.168.0.6:8333\"")
+            + HelpExampleCli("disconnectnode", "\"\" 1")
             + HelpExampleRpc("disconnectnode", "\"192.168.0.6:8333\"")
+            + HelpExampleRpc("disconnectnode", "\"\", 1")
         );
 
     if(!g_connman)
         throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
 
-    bool ret = g_connman->DisconnectNode(request.params[0].get_str());
-    if (!ret)
+    bool ret;
+    std::string address;
+
+    if (request.params[0].isStr()) {
+        address = request.params[0].get_str();
+    }
+
+    if (request.params.size() == 1) {
```

Comment:
Right, I also wondered about that.
My recommendation, also for readability, would be to structure this symmetrically e.g.
```
const UniValue &address_arg = request.params[0];
const UniValue &id_arg = request.params.size() < 2 ? NullUniValue : request.params[1];
...
if (!address_arg.IsNull() && id_arg.IsNull()) {
    /* handle kick-by-address */
    std::string address = address_arg.get_str();
} else if (!id_arg.IsNull() && address_arg.IsNull()) {
    /* handle kick-by-id */
    NodeId nodeid = (NodeId) id_arg.get_int64();
} else {
    throw JSONRPCError(RPC_INVALID_PARAMS, "Only one of adress and nodeid should be provided.");
}
```